diff --git a/deviceinfo b/deviceinfo
index e5acea3..73cf53f 100644
--- a/deviceinfo
+++ b/deviceinfo
@@ -4,8 +4,8 @@ deviceinfo_codename="mimir"
 deviceinfo_arch="aarch64"
 deviceinfo_kernel_source="https://gitlab.com/ubports/porting/reference-device-ports/halium13/volla-tablet/android_kernel_volla_mt8781.git"
 deviceinfo_kernel_source_branch="halium-13.0"
-deviceinfo_kernel_defconfig="gki_defconfig mimir.config entry_level.config halium.config"
-deviceinfo_kernel_cmdline="bootopt=64S3,32N2,64N2 systempart=/dev/mapper/system"
+deviceinfo_kernel_defconfig="gki_defconfig mimir.config entry_level.config halium.config openharmony.config"
+deviceinfo_kernel_cmdline="bootopt=64S3,32N2,64N2 systempart=/dev/mapper/system hardware=mimir ohos.boot.sn=0a20230726rpi"
 deviceinfo_kernel_clang_compile="true"
 deviceinfo_kernel_llvm_compile="true"
 deviceinfo_kernel_clang_branch="android12L-gsi"
